src/python/pants/engine/legacy/graph.py

```diff
@@ -254,6 +245,25 @@ def _inject(self, subjects):
           yielded_addresses.add(address)
           yield address
 
+  def _assert_type_is_return(self, node, state):
+    if type(state) is Return:
+      return
+
+    # NB: ResolveError means that a target was not found, which is a common user facing error.
+    # TODO Come up with a better error reporting mechanism so that we don't need this as a special case.
+    #      Possibly as part of https://github.com/pantsbuild/pants/issues/4446
+    if isinstance(state.exc, ResolveError):
+      raise AddressLookupError(str(state.exc))
+    else:
+      trace = '\n'.join(self._scheduler.trace())
+      raise AddressLookupError(
+        'Build graph construction failed for {}:\n{}'.format(node, trace))
+
+  def _assert_correct_value_type(self, state, expected_type):
```

Comment:
Hm. Maybe. In another review? I'll add another todo for #4229 here. I think cleaning up the result class is definitely related to that and it would make sense to put that check in an associated place.